samples: pmic: npm1300_one_button: Added sample (#13158)
nrfconnect/sdk-nrf: samples/pmic/native/npm1300_one_button/README.rst

.. _npm1300_one_button:

nPM1300: One button
###################

.. contents::
   :local:
   :depth: 2

The One button sample demonstrates how to support wake-up, shutdown and user interactions through
a single button connected to the nPM1300.

Requirements
************

The sample supports the following development kits:

.. table-from-sample-yaml::

The sample also requires an nPM1300 EK.

Overview
********

> The sample uses a single button for controlling the device state.

Reviewer: State here that button marked SHPHLD/RESET on EK is used.

The sample controls an LED, load switch and the PMIC power mode based on button press duration.

Wiring
******

#. Make the following connections on the nPM1300 EK:

   * Remove all existing connections.
   * Connect a suitable battery to the **J2** connector.
   * On the **P17** pin header, connect **HOST** and **LED2** pins with a jumper.
   * On the **P13** pin header, connect **RSET1** and **VSET1** pins with a jumper.
   * On the **P14** pin header, connect **RSET2** and **VSET2** pins with a jumper.
   * On the **LOAD SWITCH CONFIG** pin header, connect **VOUT2** and **LSIN2** pins with a jumper.

> 

Reviewer: I recommend we use LS1 instead, the initial EKs are using LS2 as a workaround for an issue on EK v0.9.2, so LS2 will not be available.
We can also use the onboard high load easier this way:
VOUT1 - LSIN1
HIGH - LSOUT2

Author: The code is actually using LDSW1, but the wiring info is wrong, I will update it.

#. Connect the chosen DK to the nPM1300 EK as in the following table:

   .. list-table:: nPM1300 EK connections.
      :widths: 25 25 25 25 25
      :header-rows: 1

      * - nPM1300 EK pins
        - nRF52 DK pins
        - nRF52840 DK pins
        - nRF5340 DK pins
        - nRF9160 DK pins
      * - SDA
        - P0.26
        - P0.26
        - P1.02
        - P0.30
      * - SCL
        - P0.27
        - P0.27
        - P1.03
        - P0.31

>       * - GPIO3

Reviewer: Should we state the use of GPIO3 here?

Reviewer: @aasinclair , I don't see any description of what GPIO3 is used for, maybe it's described somewhere else?

Author: I'll add it to the overview.
I would prefer not to add it in the table if possible, as it will get wider as more boards are added.

        - P0.22
        - P0.22
        - P1.04
        - P0.10
      * - VDDIO
        - VDD
        - VDD
        - VDD
        - VDD
      * - GND
        - GND
        - GND
        - GND
        - GND

Building and running
********************

.. |sample path| replace:: :file:`samples/pmic/native/npm1300_one_button`

.. include:: /includes/build_and_run.txt

Testing
*******

|test_sample|

#. |connect_kit|
#. |connect_terminal|

If the initialization was successful, the terminal displays the following message with status information:

.. code-block:: console

   PMIC device ok

Different length button presses

.. _table::
   :widths: auto

   ============  ============  =====================================================
   Duration      Log output    Outcome
   ============  ============  =====================================================
   < 1 sec       Short press   Sample flashes LED at 5Hz, LDSW 1 enabled
   1 - 5 sec     Medium press  Sample flashes LED at 1Hz, LDSW 1 disabled

>    5 - 10 sec    Long press    Sample enters ship mode provided USB is not connected

Reviewer 2: It is not clear from the user perspective what happens when "Sample enters ship mode", especially since host development kit is powered through USB.

>    5 - 10 sec    Long press    Sample enters ship mode provided USB is not connected

Reviewer: 5 - 10 sec Long press.....add at end of outcome description: Press button to wake up nPM1300 from ship mode and power up connected SoC.

>    > 10 sec      None          PMIC power cycles system

Reviewer 2: There's no code for this? Is this hardware logic in PMIC?

Also, it is not clear from the user perspective what happens when "PMIC power cycles system", especially since host development kit is powered through USB.

Author: I'll clarify this.  Yes the 10 second time is HW reset by the PMIC.

Reviewer 2: I think it would add more value to the sample if we mention this in description or in code comments.

.. _npm1300_one_button:

nPM1300: One button
###################

.. contents::
   :local:
   :depth: 2

The One button sample demonstrates how to support wake-up, shutdown and user interactions through
a single button connected to the nPM1300.

Requirements
************

The sample supports the following development kits:

.. table-from-sample-yaml::

The sample also requires an nPM1300 EK.

Overview
********

> The nPM1300 SHPHLD/RESET button controls the device state.

Reviewer 3: ```suggestion
The nPM1300 **SHPHLD/RESET** button controls the device state.
```

The sample controls an LED, load switch and nPM1300 power mode based on button press duration.

> The nPM1300 GPIO3 pin is configured as an interrupt output, and is used to signal button press
> and USB detection events to the host.

Reviewer 3: ```suggestion
The nPM1300 **GPIO3** pin is configured as an interrupt output, and is used to signal button press
and USB detection events to the host.
```

   * Remove all existing connections.
   * On the **P1** pin header, connect **VBATIN** and **VBAT** pins with a jumper.
   * On the **P13** pin header, connect **RSET1** and **VSET1** pins with a jumper.
   * On the **P14** pin header, connect **RSET2** and **VSET2** pins with a jumper.
   * On the **P15** pin header, connect **VOUT1** and **LSIN1** pins with a jumper.
   * On the **P17** pin header, connect **HOST** and **LED2** pins with a jumper.
   * Connect a suitable battery to the **J2** connector.

#. Connect the chosen DK to the nPM1300 EK as in the following table:

   .. list-table:: nPM1300 EK connections.
      :widths: 25 25 25 25 25
      :header-rows: 1

      * - nPM1300 EK pins
        - nRF52 DK pins
        - nRF52840 DK pins
        - nRF5340 DK pins
        - nRF9160 DK pins
      * - SDA
        - P0.26
        - P0.26
        - P1.02
        - P0.30
      * - SCL
        - P0.27
        - P0.27
        - P1.03
        - P0.31
      * - GPIO3
        - P0.22
        - P0.22
        - P1.12
        - P0.10
      * - VDDIO
        - VDD
        - VDD
        - VDD
        - VDD
      * - GND
        - GND
        - GND
        - GND
        - GND

> 

Reviewer 3: ```suggestion
```
Just removing an extra empty line.

Building and running
********************

.. |sample path| replace:: :file:`samples/pmic/native/npm1300_one_button`

.. include:: /includes/build_and_run.txt

Testing
*******

|test_sample|

#. |connect_kit|
#. |connect_terminal|

If the initialization was successful, the terminal displays the following message with status information:

.. code-block:: console

   PMIC device ok

> 

Reviewer 3: ```suggestion
```
Another extra empty line.

> Different length button presses

Reviewer 3: ```suggestion
Different length button presses:
```

.. _table::
   :widths: auto

   ============  ============  ======================================================
   Duration      Log output    Outcome
   ============  ============  ======================================================

>    < 1 sec       Short press   Sample flashes LED at 5Hz, LDSW 1 enabled
>    1 - 5 sec     Medium press  Sample flashes LED at 1Hz, LDSW 1 disabled

Reviewer 3: ```suggestion
   < 1 sec       Short press   Sample flashes **LED** at 5 Hz, LDSW 1 enabled
   1 - 5 sec     Medium press  Sample flashes **LED** at 1 Hz, LDSW 1 disabled
```

>    5 - 10 sec    Long press    nPM1300 enters ship mode provided USB is not connected

Reviewer 2: I think it's better to clarify USB connector (J3).

   > 10 sec      None          nPM1300 long press reset activates
   ============  ============  ======================================================

> Ship mode is the lowest power state of the nPM1300.  To exit ship mode, the SHPHLD button must be pressed.

Reviewer 2: Or USB cable attached to J3.

> Ship mode is the lowest power state of the nPM1300.  To exit ship mode, the SHPHLD button must be pressed.

Reviewer 3: ```suggestion
Ship mode is the lowest power state of the nPM1300. To exit ship mode, press the **SHPHLD** button.
```

> In a production design, this will poweroff any device which is powered from an nPM1300 buck output.

Reviewer 2: I think it's any output, not only BUCK?

> In a production design, this will poweroff any device which is powered from an nPM1300 buck output.

Reviewer 3: ```suggestion
In a production design, this will power off any device that is powered from an nPM1300 buck output.
```

> The long press reset performs a full power cycle of the nPM1300 and reset all setting to powerup defaults.

Reviewer 3: ```suggestion
The long press reset performs a full power cycle of the nPM1300 and resets all setting to powerup defaults.
```

> In a production design, this will reset any device which is powered from an nPM1300 buck output.

Reviewer 3: ```suggestion
In a production design, this will reset any device that is powered from an nPM1300 buck output.
```

When using a separately powered development kit, the application must be restarted to reconfigure

> the nPM1300 after ship mode entry or long press reset.

Reviewer 2: I think it should be "after exiting ship mode".

> When using a separately powered development kit, the application must be restarted to reconfigure
> the nPM1300 after ship mode entry or long press reset.

Reviewer 3: ```suggestion
When using a separately powered development kit, you must restart the application to reconfigure
the nPM1300 after ship mode entry or long press reset.
```
